state/application_test.go

```diff
@@ -1928,6 +1931,121 @@ func (s *ApplicationSuite) TestAddCAASUnit(c *gc.C) {
+	application := f.MakeApplication(c, &factory.ApplicationParams{
+		CharmConfig: map[string]interface{}{
+			"foo": "bar",
+		},
+		ApplicationConfig: map[string]interface{}{
+			"app foo": "app bar",
+		},
+		ApplicationConfigFields: environschema.Fields{
+			"app foo": environschema.Attr{Type: environschema.Tstring}},
+	})
+	err := application.UpdateLeaderSettings(&goodToken{}, map[string]string{
+		"leader": "true",
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	err = application.SetMetricCredentials([]byte("sekrit"))
+	c.Assert(err, jc.ErrorIsNil)
+	dbModel, err := st.Model()
+	c.Assert(err, jc.ErrorIsNil)
+	err = dbModel.SetAnnotations(application, testAnnotations)
+	c.Assert(err, jc.ErrorIsNil)
+
+	if dbModel.Type() == state.ModelTypeCAAS {
+		caasModel, err := dbModel.CAASModel()
+		c.Assert(err, jc.ErrorIsNil)
+		err = caasModel.SetPodSpec(application.ApplicationTag(), "pod spec")
+		c.Assert(err, jc.ErrorIsNil)
+		addr := network.NewScopedAddress("192.168.1.1", network.ScopeCloudLocal)
+		err = application.UpdateCloudService("provider-id", []network.Address{addr})
+		c.Assert(err, jc.ErrorIsNil)
+	}
+
+	model, err := st.Export()
+	c.Assert(err, jc.ErrorIsNil)
+
+	applications := model.Applications()
+	c.Assert(applications, gc.HasLen, 1)
+
+	app := applications[0]
+	agentTools := version.Binary{
+		Number: jujuversion.Current,
+		Arch:   arch.HostArch(),
+		Series: app.Series(),
+	}
+
+	tools, err := application.AgentTools()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(tools.Version, gc.Equals, agentTools)
+}
+
+func (s *ApplicationSuite) TestSetAgentVersion(c *gc.C) {
+	st := s.Factory.MakeModel(c, &factory.ModelParams{
+		Name: "caas-model",
+		Type: state.ModelTypeCAAS, CloudRegion: "<none>",
+		StorageProviderRegistry: factory.NilStorageProviderRegistry{}})
+	defer st.Close()
+	f := factory.NewFactory(st)
+	application := f.MakeApplication(c, &factory.ApplicationParams{
+		CharmConfig: map[string]interface{}{
+			"foo": "bar",
+		},
+		ApplicationConfig: map[string]interface{}{
+			"app foo": "app bar",
+		},
+		ApplicationConfigFields: environschema.Fields{
+			"app foo": environschema.Attr{Type: environschema.Tstring}},
+	})
+	err := application.UpdateLeaderSettings(&goodToken{}, map[string]string{
+		"leader": "true",
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	err = application.SetMetricCredentials([]byte("sekrit"))
+	c.Assert(err, jc.ErrorIsNil)
+	dbModel, err := st.Model()
+	c.Assert(err, jc.ErrorIsNil)
+	err = dbModel.SetAnnotations(application, testAnnotations)
+	c.Assert(err, jc.ErrorIsNil)
+
+	if dbModel.Type() == state.ModelTypeCAAS {
+		caasModel, err := dbModel.CAASModel()
+		c.Assert(err, jc.ErrorIsNil)
+		err = caasModel.SetPodSpec(application.ApplicationTag(), "pod spec")
+		c.Assert(err, jc.ErrorIsNil)
+		addr := network.NewScopedAddress("192.168.1.1", network.ScopeCloudLocal)
+		err = application.UpdateCloudService("provider-id", []network.Address{addr})
+		c.Assert(err, jc.ErrorIsNil)
+	}
+
+	model, err := st.Export()
```

Comment:
No need to export anything - we are just testing the get/set tools behaviour of the application persistent state.